File: cpp/975_OddEvenJump.cpp
class Solution {
public:
    int oddEvenJumps(vector<int>& A) {
        const int N = A.size();
        if (N <= 1) return N;
        
        vector<bool> odd(N, false);
        vector<bool> even(N, false);
        map<int, int> vals = {};
        
        odd[N-1] = even[N-1] = true;
        vals[A[N-1]] = N-1;
        
        for (int i = N-2; i >= 0; --i) {
            if (vals.find(A[i]) != vals.end()) {
                odd[i] = even[vals[A[i]]];
                even[i] = odd[vals[A[i]]];
            } else {
                auto lower = vals.lower_bound(A[i]);
                if (lower-- != vals.begin()) {
                    int l = lower->first;
                    even[i] = odd[vals[l]];
                }
                
                auto higher = vals.lower_bound(A[i]);
                if (higher != vals.end()) {
                    int h = higher->first;
                    odd[i] = even[vals[h]];
                }
                
            }
            vals[A[i]] = i;
        }
        
        int count = 0;
        for (bool b : odd) {
            if (b) ++count;
        }
        return count;
    }
};

File: java/975_OddEvenJump.java
class Solution {
    public int oddEvenJumps(int[] A) {
        int N = A.length;
        if (N <= 1) return N;
        boolean[] odd = new boolean[N];
        boolean[] even = new boolean[N];
        TreeMap<Integer, Integer> vals = new TreeMap<Integer, Integer>();
        
        odd[N-1] = even[N-1] = true;
        vals.put(A[N-1], N-1);
        
        for (int i = N-2; i >= 0; --i) {
            if (vals.containsKey(A[i])) { // same element so we take the val of the next transplant outcome of next possible jump
                odd[i] = even[vals.get(A[i])];
                even[i] = odd[vals.get(A[i])];
            } else {
                Integer lower = vals.lowerKey(A[i]);
                Integer higher = vals.higherKey(A[i]);
                
                if (lower != null) even[i] = odd[vals.get(lower)];
                if (higher != null) odd[i] = even[vals.get(higher)];
            }
            vals.put(A[i], i);
        }
        
        int count = 0;
        for (boolean b: odd) {
            if (b) ++count;
        }
        return count;
    }
}
